runtime/lua/vim/diagnostic.lua

```diff
@@ -2,22 +2,95 @@ local api, if_nil = vim.api, vim.F.if_nil
 
 local M = {}
 
----@enum DiagnosticSeverity
+--- @alias Diagnostic vim.Diagnostic
+--- @class vim.Diagnostic
+--- @field bufnr? integer
+--- @field lnum integer 0-indexed
+--- @field end_lnum? integer 0-indexed
+--- @field col integer 0-indexed
+--- @field end_col? integer 0-indexed
+--- @field severity? vim.diagnostic.Severity
+--- @field message string
+--- @field source? string
+--- @field code? string
+--- @field _tags? { deprecated: boolean, unnecessary: boolean}
+--- @field user_data? any arbitrary data plugins can add
+--- @field namespace? integer
+
+--- @class vim.diagnostic.Opts
```

Comment:
Why the namespacing difference (`vim.Diagnostic` versus `vim.diagnostic.x`)?

Reply:
`vim.Diagnostic` refers to the diagnostic object, `vim.diagnostic.*` is just used for any class used within `vim.diagnostic`.

I decided `vim.Diagnostic` was preferable to `vim.diagnostic.Diagnostic` since it's a common object that is used in many places.